### PR TITLE
kubelet: add initial support for cgroupv2

### PR DESCRIPTION
**Pull request by giuseppe**
/kind feature

do a conversion from the cgroups v1 limits to cgroups v2.

e.g. cpu.shares on cgroups v1 has a range of [2-262144] while the
equivalent on cgroups v2 is cpu.weight that uses a range [1-10000].

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

**Does this PR introduce a user-facing change?**:

NONE

```release-note
Support for running on a host that uses cgroups v2 unified mode
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

- [KEP]: https://github.com/kubernetes/enhancements/pull/1370

### CODE REVIEW

File: pkg/kubelet/cm/helpers_linux.go
Comment (AkihiroSuda) [inline pkg/kubelet/cm/helpers_linux.go]: can we ditch "subsystem mountpoint" concept for v2 codes?

runc and containerd are trying to eliminate all "subsystem" concept: https://github.com/opencontainers/runc/pull/2169 https://godoc.org/github.com/containerd/cgroups/v2

File: pkg/kubelet/cm/helpers_linux.go
Comment (AkihiroSuda) [inline pkg/kubelet/cm/helpers_linux.go]: `strings.Fields`

**Review by mattjmcnaughton — COMMENTED** [review]
/uncc

Looking forward to reading the KEP :) Removing myself as a reviewer while this diff is WIP.

**Comment by giuseppe** [comment]
> Looking forward to reading the KEP :) Removing myself as a reviewer while this diff is WIP.

KEP here: https://github.com/kubernetes/enhancements/pull/1370

**Comment by AkihiroSuda** [comment]
What are the remaining TODOs?

**Comment by giuseppe** [comment]
> What are the remaining TODOs?

I don't think there is more to do beside adding support for hugetlb once it lands in a kernel release (it should be in 5.6).

Beside that, it is important at this point to get the CI running on a cgroup v2 host and see what is still left to fix

**Comment by giuseppe** [comment]
tests are passing.  I think this is ready to go as soon as we agree on the KEP

**Comment by giuseppe** [comment]
KEP was approved.  The PR is ready for review

**Review by mattjmcnaughton — COMMENTED** [review]
Cool stuff!

It will take me a couple of days to give this a full review (gotta read up on cgroup v2 a bit first :) ).

File: pkg/kubelet/cm/helpers_linux.go
Comment (AkihiroSuda) [reply pkg/kubelet/cm/helpers_linux.go]: @giuseppe WDYT?

**Comment by mattjmcnaughton** [comment]
> Cool stuff!
> 
> It will take me a couple of days to give this a full review (gotta read up on cgroup v2 a bit first :) ).

@giuseppe I haven't forgotten about this :) I just haven't had a lot of time to review larger PRs recently :) Will hopefully have thoughts by end of week. Thanks for your patience.

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (dashpole) [inline pkg/kubelet/cm/cgroup_manager_linux.go]: This argument doesn't appear to be used anywhere.

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (dashpole) [inline pkg/kubelet/cm/cgroup_manager_linux.go]: nit: Can we be consistent about nesting wrt checking m.cgroupv2 and checking if inputs are non-nil?  Here we do `if CpuShares != nil`, then `if m.cgroupv2`, and below we do the reverse.  Either one is fine, but consistency would be nice.

File: pkg/kubelet/cm/helpers_linux.go
Comment (giuseppe) [reply pkg/kubelet/cm/helpers_linux.go]: is this refactoring something that could be addressed later?  I'd like to get the current version merged as soon as possible, so that we can already start using it

Comment (AkihiroSuda) [reply pkg/kubelet/cm/helpers_linux.go]: SGTM

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (giuseppe) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: sure, I've fixed it

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (giuseppe) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: thanks, dropped now

**Comment by giuseppe** [comment]
@dashpole is the new version fine?

**Comment by giuseppe** [comment]
@yujuhong @derekwaynecarr @dchen1107 could you please review this PR?

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (dashpole) [inline pkg/kubelet/cm/cgroup_manager_linux.go]: nit: since this is already cached in https://github.com/opencontainers/runc/blob/master/libcontainer/cgroups/utils.go, can we just call IsCgroup2UnifiedMode each time we need it instead of caching the result here?

Comment (giuseppe) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: good point, thanks for noticing it.

I've pushed a fixed version

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cgroup_manager_linux.go]: does this manager work on v1 and v2 hosts?

Comment (derekwaynecarr) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: i need to check the runc code, but is this abstracing legacy from unified manager?

**Review by derekwaynecarr — COMMENTED** [review]
i want to review this a little closer.

no conversion was needed for pids and hugepages, correct?

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (giuseppe) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: yes, `NewSystemdCgroupsManager()` checks for the cgroup version used:

https://github.com/opencontainers/runc/blob/master/libcontainer/cgroups/systemd/apply_systemd.go#L118-L136

If we are running on cgroup v1, it will use `cgroupsystemd.LegacyManager`

**Comment by giuseppe** [comment]
> no conversion was needed for pids and hugepages, correct?

yes correct, both pids and hugetlb don't need any conversion.  I'll need to add support for hugetlb to cAdvisor though, once Linux 5.6 is out (it is still at 5.6-rc5), support in libcontainer is already being added.

**Comment by mattjmcnaughton** [comment]
> > Cool stuff!
> > It will take me a couple of days to give this a full review (gotta read up on cgroup v2 a bit first :) ).
> 
> @giuseppe I haven't forgotten about this :) I just haven't had a lot of time to review larger PRs recently :) Will hopefully have thoughts by end of week. Thanks for your patience.

/uncc

Whoops, I didn't get to this before @derekwaynecarr did. Tbh, I don't see myself adding anything that his review wouldn't, so going to defer to him on the review :)

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (derekwaynecarr) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: thanks for pointer, and noting that i confirmed this is in the vendored version.

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cgroup_manager_linux.go]: minor nit: may be useful to move this conversion into a helper that we can then unit test in a follow-on.

**Review by derekwaynecarr — APPROVED** [review]
great to see this progress.

/approve
/lgtm

**Comment by mattjmcnaughton** [comment]
Wow, exciting stuff @giuseppe @derekwaynecarr - big thank you for all your work!

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (giuseppe) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: PR here: https://github.com/kubernetes/kubernetes/pull/89567

**Pull request merged**
